修复了python3.8及以上版本非windows系统不能监听配置的BUG

runzhi214 opened the pull request:
修复了ISSUE：https://github.com/nacos-group/nacos-sdk-python/issues/124
BUG在于：python3.8,3.9,3.10中，
- 启动multiprocessing的时候会把process_mgr和callback_tread_pool也尝试序列话
- 没有使用从mp库中引入的RLock（即pulling_lock和server_list_lock类型错误）。

runzhi214 committed:
修复py3.8+(含3.9,3.10）版本中，监听配置会报错Not Serializable的问题（非windows环境下，RLock改用mulitprocessing库的RLock. 同时，把callback_tread_pool和process_mgr标记为transient,因为没有序列化的必要）

runzhi214 committed:
Revert "修复py3.8+(含3.9,3.10）版本中，监听配置会报错Not Serializable的问题（非windows环境下，RLock改用mulitprocessing库的RLock. 同时，把callback_tread_pool和process_mgr标记为transient,因为没有序列化的必要）"

This reverts commit 05e30aacf35207fdb31f7ac528d227d21846d545.

runzhi214 committed:
修复了py3.8+(含3.8,3.9,3.10)的监听配置时报错not serializable

在非windows环境改用mp库中的RLock,同时将不需要序列话的process_mgr和callback_tread_pool标记为transient

FIX TO ISSUE https://github.com/nacos-group/nacos-sdk-python/issues/124

runzhi214 commented:
@realJackSun

chenhao26-nineteen commented:
@runzhi214 用这个方式改了，似乎可以 ,我是MacOs的系统，但是报了一个新的错误，能辛苦解答下吗
![image](https://user-images.githubusercontent.com/35129699/206129519-d91903c3-5760-42e9-890d-4b4dd822cc85.png)

Can't pickle local object 'MyTestCase.test_add_config_watch.<locals>.test_cb'

runzhi214 committed:
本地可用的test类：修复了使用Nested Class 和 Nested Function不能被pickle的问题。测试全部通过。test_list_naming_instance_online需要配合本地服务（实际已经成功）

FIX TO ISSUE https://github.com/nacos-group/nacos-sdk-python/issues/124

Info to
https://github.com/nacos-group/nacos-sdk-python/pull/125

runzhi214 commented:
> 

这个问题是由于test_client.py中使用了nested class和nested function导致的（因为pickle不允许），把他们重命名并放在测试类外部就可以通过了.

本地全部通过。其中，test_list_naming_instance_online这个测试方法需要单独测试，因为没有发送heartbeat的方法会马上死掉，在心跳结束之前可以看到测试通过。

```

class ShareFakeWatcher:
    content = None
    count = 0


def test_cb_fake_watcher(args):
    print(args)
    ShareFakeWatcher.count += 1
    ShareFakeWatcher.content = args["content"]


class ShareLongPulling:
    content = None


def cb_long_pulling(x):
    ShareLongPulling.content = x["content"]
    print(ShareLongPulling.content)

    def test_fake_watcher(self):
        d = "test"
        g = "DEFAULT_GROUP"

        cache_key = "+".join([d, g, NAMESPACE])

        client.add_config_watcher(d, g, test_cb_fake_watcher)
        client.add_config_watcher(d, g, test_cb_fake_watcher)
        client.add_config_watcher(d, g, test_cb_fake_watcher)
        time.sleep(1)
        client.notify_queue.put((cache_key, "xxx", "md51"))
        time.sleep(1)
        self.assertEqual(ShareFakeWatcher.content, "xxx")
        self.assertEqual(ShareFakeWatcher.count, 3)

        client.remove_config_watcher(d, g, test_cb_fake_watcher)
        ShareFakeWatcher.count = 0
        client.notify_queue.put((cache_key, "yyy", "md52"))
        time.sleep(1)
        self.assertEqual(ShareFakeWatcher.content, "yyy")
        self.assertEqual(ShareFakeWatcher.count, 2)

        client.remove_config_watcher(d, g, test_cb_fake_watcher, True)
        ShareFakeWatcher.count = 0
        client.notify_queue.put((cache_key, "not effective, no watchers", "md53"))
        time.sleep(1)
        self.assertEqual(ShareFakeWatcher.content, "yyy")
        self.assertEqual(ShareFakeWatcher.count, 0)

        ShareFakeWatcher.count = 0
        client.add_config_watcher(d, g, test_cb_fake_watcher)
        time.sleep(1)
        client.notify_queue.put((cache_key, "zzz", "md54"))
        time.sleep(1)
        self.assertEqual(ShareFakeWatcher.content, "zzz")
        self.assertEqual(ShareFakeWatcher.count, 1)

        ShareFakeWatcher.count = 0
        client.notify_queue.put((cache_key, "not effective, md5 no changes", "md54"))
        time.sleep(1)
        self.assertEqual(ShareFakeWatcher.content, "zzz")
        self.assertEqual(ShareFakeWatcher.count, 0)

    def test_long_pulling(self):
        client2 = nacos.NacosClient(SERVER_ADDRESSES, username=USERNAME, password=PASSWORD)
        d = "test1_pulling"
        g = "Group1"
        g2 = "Group2"

        client2.publish_config(d, g, "test2")
        client2.publish_config(d, g2, "test2")
        time.sleep(0.5)
        # test common
        client2.add_config_watcher(d, g, cb_long_pulling)
        client2.add_config_watcher(d, g2, cb_long_pulling)
        time.sleep(0.5)
        client2.publish_config(d, g, "test")
        client2.publish_config(d, g2, "test")
        time.sleep(1)
        self.assertEqual(ShareLongPulling.content, "test")
        client2.publish_config(d, g2, u"test2中文")
        time.sleep(1)
        self.assertEqual(ShareLongPulling.content, u"test2中文")
```
<img width="530" alt="截屏2022-12-07 17 28 38" src="https://user-images.githubusercontent.com/31073262/206140918-1a3818a8-ba2a-4945-8050-3dcfc80e2dcc.png">

<img width="612" alt="截屏2022-12-07 17 30 16" src="https://user-images.githubusercontent.com/31073262/206141410-6531eb7f-8ba8-42b3-b33b-188d503176e2.png">

runzhi214 commented:
test_client.py不需要被merge，这个没什么意义

shouldsee commented:
python一旦涉及到pickling和multiprocessing移植性就变得很差啊。。。

SakigamiYang commented:
那是因为3.7→3.8改了默认的多进程启动方法，3.7默认是fork，3.8默认是spawn

SakigamiYang commented:
你在那个区分 os_sys 是不是 Windows 的地方，把 Darwin 也单独拿出去好了，Darwin 的时候，也用 Thread

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/nacos-group/nacos-sdk-python?pullRequest=125) <br/>All committers have signed the CLA.
